sdk/dapp-kit/src/hooks/wallet/useSignTransactionBlock.ts

```diff
@@ -16,12 +16,19 @@ import { useCurrentAccount } from './useCurrentAccount.js';
 import { useCurrentWallet } from './useCurrentWallet.js';
 
 type UseSignTransactionBlockArgs = PartialBy<SuiSignTransactionBlockInput, 'account'>;
+
 type UseSignTransactionBlockResult = SuiSignTransactionBlockOutput;
 
+type UseSignTransactionBlockError =
```

Comment:
If you're looking at this and notice the very very similar code between all three hooks... well you'd be right 😆. I did attempt to abstract out the common functionality into a generic `useSuiFeature` hook, but some of the type interfaces are *just* different enough to make things tricky + I couldn't find a clean way to call the feature function for a given feature

<img width="1036" alt="image" src="https://github.com/MystenLabs/sui/assets/7453188/949fd5b0-9d4e-4f92-b3b1-a997809dff61">

Reply:
not sure if its worth it but this works I think:

```ts
export function useSignPersonalMessage({
	mutationKey,
	...mutationOptions
}: UseSignPersonalMessageMutationOptions = {}) {
	return useFeature(
		(wallet) => wallet.features['sui:signPersonalMessage'],
		(feature) => feature.signPersonalMessage,
		{
			mutationKey: walletMutationKeys.signPersonalMessage(mutationKey),
			...mutationOptions,
		},
	);
}

type UseFeatureOptions<Result, Options> = Omit<
	UseMutationOptions<
		Result,
		WalletFeatureNotSupportedError | WalletNoAccountSelectedError | WalletNotConnectedError | Error,
		Options,
		unknown
	>,
	'mutationFn'
>;

export function useFeature<
	Feature extends SuiFeatures[keyof SuiFeatures],
	Options extends { account?: WalletAccount },
	Result,
>(
	getFeature: (wallet: WalletWithRequiredFeatures) => Feature | undefined,
	getMethod: (feature: Feature) => (options: Options) => Promise<Result>,
	{ mutationKey, ...mutationOptions }: UseFeatureOptions<Result, Options> = {},
) {
	const currentWallet = useCurrentWallet();
	const currentAccount = useCurrentAccount();

	return useMutation({
		mutationKey,
		mutationFn: async (options) => {
			if (!currentWallet) {
				throw new WalletNotConnectedError('No wallet is connected.');
			}

			const signerAccount = options.account ?? currentAccount;
			if (!signerAccount) {
				throw new WalletNoAccountSelectedError(
					'No wallet account is selected to sign the personal message with.',
				);
			}

			const walletFeature = getFeature(currentWallet);
			if (!walletFeature) {
				throw new WalletFeatureNotSupportedError(
					"This wallet doesn't support the `signPersonalMessage` feature.",
				);
			}

			return await getMethod(walletFeature)({
				...options,
				account: signerAccount,
			});
		},
		...mutationOptions,
	});
}

```

Reply:
while I do really appreciate this and am also impressed at the same time 😆, I'm going to punt on generalizing all of the logic since there are some other follow-up tweaks that need to be made to the hook logic + to production-ize this, and I'm afraid I'm going to spend too much time on it with the other work that needs to be completed. I'll keep a note in my head though!

Reply:
yeah, totally fair, I just have a hard time not trying to figure something out when the challenge is there.  I think re-implementing some of this a few times across hooks is totally fine